### PR TITLE
Add file class based inference API for diarization

### PR DESCRIPTION
**Pull request by SeanNaren**
# What does this PR do?

Supports:

```python
from nemo.collections.asr.models import NeuralDiarizer

diarizer = NeuralDiarizer.from_pretrained(model_name='diar_msdd_telephonic')

annotation = diarizer("an4_diarize_test.wav")

print(annotation.to_rttm())
```

User can change the device:

```python
import torch
from nemo.collections.asr.models import NeuralDiarizer

diarizer = NeuralDiarizer.from_pretrained(model_name='diar_msdd_telephonic')
diarizer.to(torch.device('cuda'))
```

Supports passing the maximum number of speakers, or the actual number of speakers in a file (oracle):

```python
annotation = diarizer("an4_diarize_test.wav", num_speakers=2)

annotation = diarizer("an4_diarize_test.wav", max_speakers=4)
```

Relies on a temporary directory, which can be specified if preferred:

```python
annotation = diarizer("an4_diarize_test.wav", out_dir="my_temp/")
```

List of available models:

```python
from nemo.collections.asr.models.msdd_models import NeuralDiarizer

print(NeuralDiarizer.list_available_models())
"""
PretrainedModelInfo(
	pretrained_model_name=diar_msdd_telephonic,
	description=For details about this model, please visit https://ngc.nvidia.com/catalog/models/nvidia:nemo:diar_msdd_telephonic,
location=https://api.ngc.nvidia.com/v2/models/nvidia/nemo/diar_msdd_telephonic/versions/1.0.0/files/diar_msdd_telephonic.nemo
)
"""
```

### TODO
* Add tests, ensure that config options are passed correctly and updated correctly. The config logic is quite messy due to mutable objects, but this should be saved for a larger refactor.
* Hopefully in the future, `NeuralDiarizer` will be a NeMo PT model file too, save for a later refactor

**Collection**: ASR, Speaker Tasks

# Changelog 
- Add file class based inference API for Diarization.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

**Review by nithinraok — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/models/msdd_models.py
Comment (SeanNaren) [reply nemo/collections/asr/models/msdd_models.py:1429]: No idea, if so then this should be set via the `domain` flag, will wait for @tango4j to suggest.

File: nemo/collections/asr/models/msdd_models.py
Comment (SeanNaren) [reply nemo/collections/asr/models/msdd_models.py:1432]: unsure exactly what is against the convention; specifying a default model?

File: nemo/collections/asr/models/msdd_models.py
Comment (SeanNaren) [reply nemo/collections/asr/models/msdd_models.py:1456]: thats quite an ugly name; how about just calling it `input_path`? I went for `inference_input` in case we wanted to just support loading a manifest_filepath.

File: nemo/collections/asr/models/msdd_models.py
Comment (SeanNaren) [reply nemo/collections/asr/models/msdd_models.py:1458]: It made me question if we should support multiple files at all. The issue with this is it just makes the API a bit confusing; I think it just makes more logical sense to run diarization multiple times IMO from a user perspective. Can be swayed due to the optimization though.

File: nemo/collections/asr/models/msdd_models.py
Comment (nithinraok) [reply nemo/collections/asr/models/msdd_models.py:1432]: I meant domain.

File: nemo/collections/asr/models/msdd_models.py
Comment (nithinraok) [reply nemo/collections/asr/models/msdd_models.py:1458]: Agree. Lets make it to support only one file inference and if users would like to do on multiple files, then they can just run the example script.

File: nemo/collections/asr/models/msdd_models.py
Comment (tango4j) [reply nemo/collections/asr/models/msdd_models.py:1429]: Why are you loading NeuralInferenceConfig to MSDD class? 
And if it is model name, `diar_msdd_telephonic` should be the name.

File: nemo/collections/asr/models/msdd_models.py
Comment (tango4j) [reply nemo/collections/asr/models/msdd_models.py:1458]: +1 Agreed on one-file inference this time.

File: nemo/collections/asr/models/configs/msdd_config.py
Comment (titu1994) [inline nemo/collections/asr/models/configs/msdd_config.py:21]: Let's use data class please instead of nested dicts

File: nemo/collections/asr/models/msdd_models.py
Comment (titu1994) [inline nemo/collections/asr/models/msdd_models.py:29]: Import guard unless it's a required dependency

File: nemo/collections/asr/models/msdd_models.py
Comment (titu1994) [reply nemo/collections/asr/models/msdd_models.py:1432]: Signature of from pretrained is supposed to be consistent across nemov- extra args like domain etc usually are not supposed to be used. Also it's not device=ste but map_location=str for Pytorch API. 

If domain is needed, since this is custom code anyway and only mimics from_pretrained, I suppose there's no choice but to keep it. But ideally I would incorporate it directly into your from pretrained name itself - for example stt_en_conformer_ctc_large_ls is completely descriptive - speech to text model, English language, Conformer architecture, CTC loss, Large model size, Librispeech dataset

File: nemo/collections/asr/models/msdd_models.py
Comment (titu1994) [reply nemo/collections/asr/models/msdd_models.py:1456]: Path2audio_files will be deprecated with future release anyway, it's fine to change the name here. But not generic input_path - audio_filepaths, audio_paths or some variation should be used. If more types of data are used, then split the args

File: nemo/collections/asr/models/msdd_models.py
Comment (titu1994) [reply nemo/collections/asr/models/msdd_models.py:1458]: How costly is it to run the 3-4 model N times vs calling them exactly N/BS time ? I would go for multiple file support but it's your domain so I'll defer to you.

File: nemo/collections/asr/models/msdd_models.py
Comment (nithinraok) [reply nemo/collections/asr/models/msdd_models.py:1456]: `audio_filepath` ? since we do inference of only one file now.

File: nemo/collections/asr/models/msdd_models.py
Comment (SeanNaren) [reply nemo/collections/asr/models/msdd_models.py:1432]: Agree that `domain` is out the norm.

Also agreement that the parameters should actually be stored with the model, however because this is still evolving (and hopefully in the future unnecessary) we still need to set these parameters somehow. Any alternatives to passing a string for picking the domain specific parameters? @tango4j @nithinraok @titu1994

File: nemo/collections/asr/models/msdd_models.py
Comment (SeanNaren) [reply nemo/collections/asr/models/msdd_models.py:1458]: Good question, but let's just support a single file for now, would be easy to add support in a separate iteration.

File: nemo/collections/asr/models/clustering_diarizer.py
Comment (SeanNaren) [reply nemo/collections/asr/models/clustering_diarizer.py:90]: Agree, might see how far we can get by just inheriting from `torch.nn.Module`

File: nemo/collections/asr/models/msdd_models.py
Comment (tango4j) [reply nemo/collections/asr/models/msdd_models.py:1458]: I get so many comments about long-form audio diarization. Current diarization implementation maxes out batch size for embedding extraction (for multiple files) then clustering is done for single files since 1h audio can be 3G GPU mem and 2h can be 10G GPU mem. So I think processing one-by-one is not a bad decision for lay-people oriented API.

File: nemo/collections/asr/models/msdd_models.py
Comment (nithinraok) [reply nemo/collections/asr/models/msdd_models.py:1432]: My view is since we are passing model names then domain doesn;t matter as those models define the type of domain, (like msdd telephonic) also msdd model itself have speaker_model and configs of multi-scale (as msdd is trained on those scale types)  and those parameters are mostly model specific. So lets keep it based on general model [config](https://github.com/NVIDIA/NeMo/blob/main/examples/speaker_tasks/diarization/conf/inference/diar_infer_general.yaml). But then general config doesn;t work well with telephonic msdd? (@tango4j ) 

So my main question is: If msdd model is telephonic, domain doesn;t matter as its only tuned for telephonic audio, and setting domain externally might actually hurt model performance as msdd model for meeting isn;t released yet. 
Does this support offline clustering without msdd?

Comment (tango4j) [reply nemo/collections/asr/models/msdd_models.py:1432]: @nithinraok I haven't trained general model config yet, because of the VAD issues for the overall diarization performance. Currently it can only use "diar_telephonic_msdd" only. Can't use general model config now. 

Like @nithinraok said, currently, domain argument cannot be doing anything. Until I upload models for other domains.

**Comment by tango4j** [comment]
I resolved the conflict in speaker_utils.py and tested with the script you provided.

File: nemo/collections/asr/models/msdd_models.py
Comment (nithinraok) [reply nemo/collections/asr/models/msdd_models.py:1432]: Taejin, have you checked performance of msdd_telephonic on clean meeting type of audio? If its bad then we need to support offline clustering without overlap detection for good performance wrt domain. 
https://github.com/NVIDIA/NeMo/blob/4804f314116cac841539aedcceec2a55e0a5a412/nemo/collections/asr/models/clustering_diarizer.py#L390 should work. So we should think of a parent class that can subclass both offline and neural diarizers to be able to diarize incoming audio.

File: nemo/collections/asr/models/configs/msdd_config.py
Comment (tango4j) [reply nemo/collections/asr/models/configs/msdd_config.py:21]: @SeanNaren Since we already made default parameters with classes, can we just use yaml file from the git?

File: nemo/collections/asr/models/msdd_models.py
Comment (tango4j) [reply nemo/collections/asr/models/msdd_models.py:1432]: @nithinraok let's focus on building this API first. I will work on releasing general model and meeting model. 

Also if we tune for meeting domain, then it is also bad for NIST-SRE-2K type of dataset too. There should be a general VAD, general scale settings and everything.

Comment (nithinraok) [reply nemo/collections/asr/models/msdd_models.py:1432]: yes, no hurry on releasing MSDD models, but I am thinking if we should support offline speaker diarization without overlap for non-telephonic incoming audio (general). That will make this API complete and usable without restriction on incoming audio. 

yes, as you said we can do clustering diarizer setup for time being. cc: @SeanNaren

Comment (tango4j) [reply nemo/collections/asr/models/msdd_models.py:1432]: @nithinraok Letting people choose clustering/MSDD is too much of information and makes us lose the point of making one-line API. So let's use MSDD only when domain='telephonic' and put clustering diarizer instead for other domains. I will publish diar_msdd_meeting shortly, then switch it to support OvlDetection.

Comment (titu1994) [reply nemo/collections/asr/models/msdd_models.py:1432]: While I agree, there's no need for that level of rigidity. ASR handles this with "change decoding strategy" that is a specialized class that is a middle layer betwee model output and metric calculation - taking up the tasks of decoding the models output from and into various forms. 

For example while models output is always logits or logprobs, we may want to do beam or greedy decoding so we make the decoding logical flexible and dynamic (changeable at runtime). You could take a similar approach in the future

Comment (SeanNaren) [reply nemo/collections/asr/models/msdd_models.py]: For a first iteration, I've removed domain. We can just specify the model(s) to load

File: nemo/collections/asr/models/msdd_models.py
Comment (SeanNaren) [reply nemo/collections/asr/models/msdd_models.py]: Removed multi-file support

File: nemo/collections/asr/models/configs/msdd_config.py
Comment (SeanNaren) [reply nemo/collections/asr/models/configs/msdd_config.py]: Removed the need for this as we do not have domain anymore!

File: nemo/collections/asr/models/msdd_models.py
Comment (SeanNaren) [reply nemo/collections/asr/models/msdd_models.py:29]: we already import `DiarizationErrorRate` from pyannote a line below, is an import guard necessary?

File: nemo/collections/asr/models/clustering_diarizer.py
Comment (SeanNaren) [reply nemo/collections/asr/models/clustering_diarizer.py:90]: I've mimicked `device` and `to` as now described in the README. turns out it will require quite a large refactor to make it fully support `device` and `to` as a torch.nn.Module. We should save this for a separate PR.

**Comment by SeanNaren** [comment]
Thanks guys. I have made modifications as per requested to the code.

> I think it is required to add a simple CI-test for this too? I guess it won't be that difficult.

Yep, once we're happy with the API and stuff, can start to add tests for functionality.

The current code seems to be a one run setup; the code isn't setup for `diarize` to be called multiple times. It instantiates all the models again in the pipeline.

I'll be modifying this behaviour.

File: nemo/collections/asr/models/clustering_diarizer.py
Comment (SeanNaren) [reply nemo/collections/asr/models/clustering_diarizer.py:90]: Managed to get this working via LightningModule.

**Comment by SeanNaren** [comment]
High level updates:

* Tried to reduce logging verbosity as much as I can. This is controlled by the `verbose` flag. This will reduce info logging to a minimum. Much more readable, however I'd like to reduce the amount of spam even further
* `NeuralDiarizer` is now based on a `LightningModule` allowing us to easily swap between devices
* It supports multiple files now. Not the cleanest as we're updating a config data-class to do so, but works for now

File: nemo/collections/asr/models/msdd_models.py
Comment (nithinraok) [reply nemo/collections/asr/models/msdd_models.py:29]: not necessary, pyannote is part of our requirements

**Comment by tango4j** [comment]
I have sent you the branch `feat/inference_diar_add` with the necessary changes to make sure these things: 
- we only use MSDD for telephonic and the rest is clustering diarizer
- download yaml file from `raw.githubusercontent.com` for automate the yaml file update for API.

File: nemo/collections/asr/models/msdd_models.py
Comment (nithinraok) [inline nemo/collections/asr/models/msdd_models.py:997]: why is audio_rttm_map is removed from here?

Comment (SeanNaren) [reply nemo/collections/asr/models/msdd_models.py:997]: This cannot be defined within the `__init__` as the manifest filepath is not set till `diarize` is called.

It has been moved to here: https://github.com/NVIDIA/NeMo/pull/5945/files/aff83a2532459027ad09c5e68a1b4e17b1801df4#diff-83cbff1d089e98d8c1551c3214d038a32ac03fc3fe6f454e7c0810af149a02d8R1416

**Comment by SeanNaren** [comment]
After offline discussion the high level changes are as below:

* Move telephonic configuration to the data-class. This will be our new “default” for the neural diarizer (especially considering the pre-trained model is labelled telephonic). Remove the telephonic yaml file
* The general and meeting yaml files will inherit parameters from the data-class config, just like ASR does. We remove all the parameters that are not changed as they will be inherited from the data-class when loaded
 
When changes to the defaults are made, they are made in the data-class config file (the python data-class). When changes regarding meeting or general are made, they are made in the yaml files. This means parameters are not duplicated, but there is two different places to update (which I think is fine as maintainers).

One note to make, is that our neural diarizer model is telephonic. If we release further adapted models to different domains (which is not certain yet) we will need to come up with a mechanism via the API to enable this.

**Comment by tango4j** [comment]
> After offline discussion the high level changes are as below:
> 
> * Move telephonic configuration to the data-class. This will be our new “default” for the neural diarizer (especially considering the pre-trained model is labelled telephonic). Remove the telephonic yaml file
> * The general and meeting yaml files will inherit parameters from the data-class config, just like ASR does. We remove all the parameters that are not changed as they will be inherited from the data-class when loaded
> 
> When changes to the defaults are made, they are made in the data-class config file (the python data-class). When changes regarding meeting or general are made, they are made in the yaml files. This means parameters are not duplicated, but there is two different places to update (which I think is fine as maintainers, as the config file is mainly to try different configurations out).
> 
> One note to make, is that our neural diarizer model is telephonic. If we release further adapted models to different domains (which is not certain yet) we will need to come up with a mechanism via the API to enable this.

I recognize that making data class is necessary but editing diarization yaml file this way is too confusing. Even I am having a hard time grasping what is happening with the configuration. Also, setting telephonic as a default configuration for the whole diarization pipeline is also confusing.

Can we just duplicate the configuration (for now, it is telephonic) and just revert all yaml files? This is too much of change on the original diarization pipeline, which is only focusing on API purpose.

File: nemo/collections/asr/models/msdd_models.py
Comment (SeanNaren) [inline nemo/collections/asr/models/msdd_models.py:1193]:
```suggestion
        return None
```

**Comment by SeanNaren** [comment]
Thanks @tango4j I have updated the PR based on your suggestions.

Would we be able to review and merge this now?

File: nemo/collections/asr/models/clustering_diarizer.py
Comment (tango4j) [inline nemo/collections/asr/models/clustering_diarizer.py:457]: Is this necessary change? If RTTM is not provided, it does not evaluate and skip.

Also, I would prefer putting return to `else` case although it does not change anything.

**Review by tango4j — CHANGES_REQUESTED** [review]
Couple of points to be addressed.
Also it would be great if we can remove the scale-specific tqdm in clustering diarizer.
It is overkill to show it.

File: nemo/collections/asr/models/clustering_diarizer.py
Comment (SeanNaren) [reply nemo/collections/asr/models/clustering_diarizer.py:457]: ah you are right, this does nothing! reverted this.

**Comment by SeanNaren** [comment]
Thanks @tango4j everything addressed.

I've reduced the TQDM logging by introducing verbosity to those functions. Everything left is just general NeMo logging, which also needs to definitely be reduced (but in a different PR I believe).

**Review by tango4j — DISMISSED** [review]
I will do PRs when I need to update the configs later.
Thank you for your effort on building this.

File: nemo/collections/asr/models/clustering_diarizer.py
Comment (titu1994) [inline nemo/collections/asr/models/clustering_diarizer.py:75]: torch.nn.Module as first in inheritance list

File: nemo/collections/asr/models/clustering_diarizer.py
Comment (titu1994) [inline nemo/collections/asr/models/clustering_diarizer.py:120]: Maybe use nemo.utils.logging?

File: nemo/collections/asr/models/clustering_diarizer.py
Comment (titu1994) [inline nemo/collections/asr/models/clustering_diarizer.py:412]: Add some checks before removing dirs

File: nemo/collections/asr/models/msdd_models.py
Comment (titu1994) [inline nemo/collections/asr/models/msdd_models.py:1458]: Very cool

File: examples/speaker_tasks/diarization/conf/inference/diar_infer_general.yaml
Comment (nithinraok) [inline examples/speaker_tasks/diarization/conf/inference/diar_infer_general.yaml:15]: Is this parameter required? We are evaluating based on if RTTM files are passed or not

File: examples/speaker_tasks/diarization/conf/inference/diar_infer_meeting.yaml
Comment (nithinraok) [inline examples/speaker_tasks/diarization/conf/inference/diar_infer_meeting.yaml:13]: I didn;t notice this last time. Is `map_location`  different from `device`?

Comment (SeanNaren) [reply examples/speaker_tasks/diarization/conf/inference/diar_infer_meeting.yaml:13]: Yes, `map_location` is used to determine if the model weights device. This just exposes the default to ensure when CPU is chosen, we do not load on the GPU.

File: examples/speaker_tasks/diarization/conf/inference/diar_infer_general.yaml
Comment (SeanNaren) [reply examples/speaker_tasks/diarization/conf/inference/diar_infer_general.yaml]: Thanks removed this and pass the verbose flag to the logging warning.

File: examples/speaker_tasks/diarization/conf/inference/diar_infer_meeting.yaml
Comment (nithinraok) [reply examples/speaker_tasks/diarization/conf/inference/diar_infer_meeting.yaml:13]: Then I don;t see use of device now. Lets keep only one and remove other

Comment (tango4j) [reply examples/speaker_tasks/diarization/conf/inference/diar_infer_meeting.yaml:13]: Didn't you already add `device` for this reason ? Since inference is only supporting 1 GPU inference, I don't see the need of map_location while we have `device`.

Comment (tango4j) [reply examples/speaker_tasks/diarization/conf/inference/diar_infer_meeting.yaml:13]: @SeanNaren Could you address this question? It will be great to make this clear - why do we need `map_location` for inference

Comment (SeanNaren) [reply examples/speaker_tasks/diarization/conf/inference/diar_infer_meeting.yaml]: Addressed issues, for the config file we rely on `device` which is required for the hydra runner as device needs to be specified beforehand.

Regarding why map location is needed, this is just imitating the PyTorch standard API:

```python
# load the model onto CUDA:1
diarizer = NeuralDiarizer.from_pretrained(model_name='diar_msdd_telephonic', map_location='cuda:1')
diarizer.to(torch.device('cuda'))

annotation = diarizer("an4_diarize_test.wav", max_speakers=4, out_dir='diar/')
```

I think I've addressed this issue though! there is now just one internal variable; device.

**Review by nithinraok — APPROVED** [review]
Great work Sean.

**Pull request merged**
